util/licence-checker.py

```diff
@@ -0,0 +1,357 @@
+
+    def excluded(self, path, reason):
+        rel_path = path.relative_to(self.base_dir)
+        logging.debug("%s: EXCLUDED: %s", str(rel_path), reason)
+        self.excluded_count += 1
+
+
+    def any_failed(self):
+        return self.failed_count > 0
+
+    def display_nicely(self):
+        return """
+        Results:     Files
+        ------------------
+        Passed:   {:>8d}
+        Failed:   {:>8d}
+        Skipped:  {:>8d}
+        Excluded: {:>8d}
+        ------------------
+        Total:    {:>8d}
+        """.format(self.passed_count, self.failed_count,
+                              self.skipped_count, self.excluded_count, self.total_count)
+
+def matches_exclude_pattern(config, file_path):
+    rel_path = str(file_path.relative_to(config.base_dir))
+    for exclude_pattern in config.exclude_paths:
+        if fnmatch.fnmatch(rel_path, exclude_pattern):
+            return True
+    return False
+
+
+def check_paths(config, git_paths):
+    results = ResultsTracker(config.base_dir)
+
+    for filepath in git_find_all_file_paths(config.base_dir, git_paths):
+        # Skip symlinks (with message)
+        if filepath.is_symlink():
+            results.excluded(filepath, "File is a symlink")
+            continue
+
+        # Skip non-file
+        if not filepath.is_file():
+            continue
+
+        # Skip exclude patterns
+        if matches_exclude_pattern(config, filepath):
+            results.excluded(filepath, "Path matches exclude pattern")
+            continue
+
+        check_file_for_licence(config.licence, results, filepath)
+
+    return results
+
+
+def check_file_for_licence(licence, results, filepath):
+    comment_style = detect_comment_char(filepath.name)
+
+    if comment_style is None:
+        results.skipped(filepath, "Unknown comment style")
+        return
+
+    if filepath.stat().st_size == 0:
+        results.skipped(filepath, "Empty file")
+        return
+
+    def next_line(file, line_no):
+        return (next(file).rstrip(), line_no + 1)
+
+    with filepath.open() as f:
+        licence_assumed_start = None
+
+        # Get first line
+        try:
+            line, line_no = next_line(f, 0)
+        except StopIteration:
+            results.failed(filepath, 1, "Empty file")
+            return
+
+        licence_search_marker = comment_style.search_line(licence.first_word)
+
+        # Check first line against the first word of licence, or against a
+        # possible different first line.
+        if not line.startswith(licence_search_marker):
+            if not line.startswith(comment_style.first_line_prefix):
+                results.failed(filepath, line_no,
+                               "File does not start with comment")
+                return
+
+            try:
+                line, line_no = next_line(f, line_no)
+            except StopIteration:
+                results.failed(filepath, line_no, "Reached end of file before finding licence")
```

Comment:
Should this have a slightly different message to make it clear that we didn't get past the 1st line?

Reply:
The line number is printed with the filename, which hopefully is enough of a clue?

Reply:
Ah, good point. Sounds sensible enough to me. I generally try to make "subtle" changes to error messages for this sort of thing to make debugging my script easier, but that's arguably a bit hacky!

Reply:
I worked pretty hard on the line number stuff so that I had that to help me when I was debugging, but then my terminal will turn `<filename>:<number>` into a clickable link that takes me exactly there, which made debugging quite a bit easier.